framework/src/problems/ReferenceResidualProblem.C

```diff
@@ -337,33 +360,46 @@ ReferenceResidualProblem::updateReferenceResidual()
   System & s = nonlinear_sys.system();
   auto & as = aux_sys.sys();
 
-  for (unsigned int i = 0; i < _group_resid.size(); ++i)
-  {
-    _group_resid[i] = 0.0;
-    _group_output_resid[i] = 0.0;
-    _group_ref_resid[i] = 0.0;
-  }
+  std::fill(_group_resid.begin(), _group_resid.end(), 0.0);
+  std::fill(_group_output_resid.begin(), _group_output_resid.end(), 0.0);
+  if (_local_norm)
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 1.0);
+  else
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 0.0);
 
   for (unsigned int i = 0; i < _soln_vars.size(); ++i)
   {
-    const auto resid =
-        Utility::pow<2>(s.calculate_norm(nonlinear_sys.RHS(), _soln_vars[i], DISCRETE_L2));
+    Real resid = 0.0;
     const auto group = _variable_group_num_index[i];
-    _group_resid[group] += _converge_on_var[i] ? resid : 0;
-    _group_output_resid[group] += resid;
-    if (_reference_vector)
+    if (_local_norm)
+    {
+      mooseAssert(nonlinear_sys.RHS().size() == (*_reference_vector).size(),
+                  "Sizes of nonlinear RHS and reference vector should be the same.");
+      mooseAssert((*_reference_vector).size(), "Reference vector must be provided.");
+      auto div = nonlinear_sys.RHS().clone();
+      *div /= *_reference_vector;
+      resid = Utility::pow<2>(s.calculate_norm(*div, _soln_vars[i], DISCRETE_L_INF));
```

Comment:
calculate_norm does some magic masking for each variable so it's worth keeping it here rather than doing a custom loop.